### ISSUE TITLE
ACI provider prematurely reports pods as ready when they, in fact, are unreachable

### ISSUE THREAD

**Issue by krancour** [virtual-kubelet/virtual-kubelet#448]
---

### Environment summary

Provider (e.g. ACI, AWS Fargate, Hyper)

ACI

Version (e.g. 0.1, 0.2-beta)

v0.7.4

K8s Master Info (e.g. AKS, ACS, Bare Metal, EKS)

AKS (k8s v1.11.5)

Install Method (e.g. Helm Chart, )

Helm chart-- the virtual-kubelet-for-aks chart, specifically

### Issue Details

Consistently, I can observe that pods scheduled to VK remain unreachable for sometimes as long as approximately two minutes _after_ they've been assigned IPs, gone into a ready condition, and even been selected as endpoints for a service.

### Repo Steps

Apply this manifest to your cluster:

```
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  type: LoadBalancer
  ports:
  - port: 80
    targetPort: http
  selector:
    app: nginx
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 0
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - name: http 
          containerPort: 80
        livenessProbe:
          httpGet:
            path: /
            port: 80
        readinessProbe:
          httpGet:
            path: /
            port: 80
      nodeSelector:
        kubernetes.io/role: agent
        beta.kubernetes.io/os: linux
        type: virtual-kubelet
      tolerations:
      - key: virtual-kubelet.io/provider
        operator: Exists
      - key: azure.com/aci
        effect: NoSchedule
```

1. Wait for the `nginx` service to get an external IP.

    ```
    watch kubectl get svc nginx
    ```

1. Watch `nginx` endpoints.

    In a separate terminal:

    ```
    watch kubectl get endpoints nginx
    ```

    (There will be none, initially.)

1. Watch nginx pods.

    In a separate terminal:

    ```
    watch kubectl get pods -l app=nginx
    ```

1. Scale the `nginx` deployment to one replica.

    In a separate terminal:

    ```
    kubectl scale deployment nginx --replicas 1
    ```

1. Wait for the pod to appear ready. (See no. 3)

1. Wait for the pod's IP to appear among the `nginx` endpoints. (See no. 2)

1. Curl the service using its external IP.

    ```
    curl -s -o /dev/null -w "%{http_code}" http://<service external IP>
    ```

    Observe that either:

      a. A successful response will take quite some time

      OR

      b. The request will time out with an "http response code" (not really) of 000, which indicates a dialing error.

    Repeated attempts to curl the IP will eventually be successful.

1. Scale the deployment back down to zero:

    ```
    kubectl scale deployment nginx --replicas 0
    ```

1. Scale the deployment back up to one:

    ```
    kubectl scale deployment nginx --replicas 0
    ```

1. Wait for the pod to appear ready. (See no. 3)

1. Wait for the pod's IP to appear among the `nginx` endpoints. (See no. 2)

1. Curl the pod _directly_ (it has a public IP, after all).

    ```
    curl -s -o /dev/null -w "%{http_code}" http://<pod's public IP>
    ```

    Observe that either:

      a. A successful response will take quite some time

      OR

      b. The request will time out with an "http response code" (not really) of 000, which indicates a dialing error.

    Repeated attempts to curl the IP will eventually be successful.

All of this would strongly indicate that VK (or ACI itself?) prematurely reports pods to be "ready," when, in fact, they're not yet accessible via their assigned IPs.

**Comment by krancour** [comment]
Cross-referencing https://github.com/osiris-org/osiris/issues/152

**Comment by pires** [comment]
This seems a provider issue and not a VK issue.

**Comment by krancour** [comment]
> This seems a provider issue and not a VK issue.

That seems quite possible. I don't know the VK implementation well-enough to assume that. I figured this would a good starting point for getting this addressed.

**Comment by pires** [comment]
@krancour thanks for clarifying. So yeah, it's an ACI provider issue as you reported in the issue description.

**Comment by srrengar** [comment]
@krancour I am unable to repro this - I can hit the endpoint successfully within seconds after the pod is up. Could you see if you are still facing this issue? 

I did install the connector with the virtual nodes add on, not helm but I'm not sure if that would matter?

**Comment by srrengar** [comment]
with more investigation I found out that the liveness and/or readiness probes cause the container in the pod to restart repeatedly (I confirmed this with another pod as well) and these aren't supported yet with the ACI provider, so maybe that was the issue. @rbitia to add more if needed.

**Comment by krancour** [comment]
> I did install the connector with the virtual nodes add on, not helm but I'm not sure if that would matter?

It's been some time since I looked at this, but iirc, there may actually be differences between the two. @rbitia may know more.

> with more investigation I found out that the liveness and/or readiness probes cause the container in the pod to restart repeatedly

Yes. That is the behavior I also encountered with the add-on (not the VK Helm chart). I seem to recall having opened a separate issue for that, internally.

i.e. Two separate options for installation-- two separate issues. This one is specific to VK as installed via Helm.
